src/Estimators/src/InvariantEKFBaseEstimator.cpp

```diff
@@ -0,0 +1,918 @@
+    }
+
+    // should we handle removing old contacts and adding new contacts? or let it be as it is.
+    return true;
+}
+
+
+bool InvariantEKFBaseEstimator::Impl::propagateStates(const FloatingBaseEstimators::Measurements& meas,
+                                                      const double& dt,
+                                                      const Eigen::Vector3d& g,
+                                                      const iDynTree::Transform& IMU_H_RF,
+                                                      const iDynTree::Transform& IMU_H_LF,
+                                                      FloatingBaseEstimators::InternalState& X)
+{
+    Eigen::Vector3d acc_unbiased = meas.acc - X.accelerometerBias;
+    Eigen::Vector3d gyro_unbiased = meas.gyro - X.gyroscopeBias;
+
+    Eigen::Matrix3d R = X.imuOrientation.toRotationMatrix();
+    Eigen::Vector3d v = X.imuLinearVelocity;
+    Eigen::Vector3d p = X.imuPosition;
+    Eigen::Vector3d acc = (R*acc_unbiased) + g;
+
+    manif::SO3Tangentd omega_skew_dt(gyro_unbiased*dt);
+    Eigen::Matrix3d R_pred = R*(omega_skew_dt.exp().rotation());
+
+    X.imuOrientation = Eigen::Quaterniond(R_pred);
+    X.imuLinearVelocity = v + acc*dt;
+    X.imuPosition = p + (v*dt) + (acc*(0.5*dt*dt));
+
+    if (!meas.lfInContact)
+    {
+        X.lContactFramePosition = X.imuPosition + (R*iDynTree::toEigen(IMU_H_LF.getPosition()));
+    }
+
+    if (!meas.rfInContact)
+    {
+        X.rContactFramePosition = X.imuPosition + (R*iDynTree::toEigen(IMU_H_RF.getPosition()));
+    }
+
+    return true;
+}
+
+bool InvariantEKFBaseEstimator::Impl::updateStates(const Eigen::VectorXd& obs,
+                                                   const Eigen::MatrixXd measModelJacobian,
+                                                   const Eigen::MatrixXd& measNoiseVar,
+                                                   const Eigen::MatrixXd& auxMat,
+                                                   FloatingBaseEstimators::InternalState& state,
+                                                   Eigen::MatrixXd& P)
+{
+    if (measModelJacobian.cols() != P.rows())
+    {
+        std::cerr << "[InvariantEKFBaseEstimator::updateStates] Measurement model Jacobian size mismatch" << std::endl;
+        return false;
+    }
+
+    if (measModelJacobian.rows() != measNoiseVar.rows())
+    {
+        std::cerr << "[InvariantEKFBaseEstimator::updateStates] Measurement noise covariance matrix size mismatch" << std::endl;
+        return false;
+    }
+
+    bool estimateBias;
+    if (P.rows() == m_vecSizeWBias)
+    {
+        estimateBias = true;
+    }
+    else
+    {
+        estimateBias = false;
+    }
+
+    Eigen::MatrixXd PHT = P*measModelJacobian.transpose();
+    Eigen::MatrixXd S = measModelJacobian*PHT + measNoiseVar;
+    Eigen::MatrixXd K = PHT*(S.inverse());
```

Comment:
As before, probably we may avoid having this continuous memory allocation.

Reply:
I'd prefer to keep this part as it is since this also depends on the number of feet in contact with the ground which seems to be changing "almost continuously". This means anyways we might have to do an almost continuous resizing of these matrices. Instead for the above comment, I made the matrices `Fc`, `Qc` and `Lc` as class members to avoid continuous memory allocation, since their dimensions would only change at very rare events of toggling the bias estimation.

What do you think @GiulioRomualdi?

Reply:
To be honest, I am with @GiulioRomualdi on this. It is always better to avoid memory allocations, especially in pieces of code that are called at every control iteration. In principle, these should have no memory allocation at all. 
It is acceptable to have some resize from time to time, also considering that the container may have some capacity (hence memory may not be allocated also in case of resizes). 
If the number of members blows up, then we can discuss alternative solutions.

Also, I have noticed that in the code below there are several temporary matrices (``X``, ``BigX``, ``innovation``, ``delta``, ``dX``, ``IminusKH``). Also this should be avoided as much as possible.

Reply:
I've partially done this in https://github.com/dic-iit/bipedal-locomotion-framework/pull/85/commits/6abd67857f238dc60818b25e497be228b62ab230. Let me know if this is in line with what is suggested. However, about the resizing of the matrices related to the measurement model/updates, it would require a bit more time to do it in a cleaner way.


I also completely agree with what is suggested. In fact, I had an earlier design in the element repo which considered these things. But not sure why I changed my mind to do things in a different way here.

Reply:
``Eigen`` automatically resizes the matrix when using the assignment operator. So you don't need to deal with the explicit resize for a first iteration I think. Then, later on, we can understand if it should be better to resize these matrices beforehand in a specific place.

Reply:
Fixed in https://github.com/dic-iit/bipedal-locomotion-framework/pull/85/commits/d311aa2da5cc8750123dce7076d0c041f93e3e08